src/drug_discovery.py

```diff
@@ -0,0 +1,712 @@
+      - protein (list of protein files)
+      - ABFE (ligand, protein, step, job_id, output, results, delta_g)
+      - RBFE (ligand1, ligand2, protein, step, job_id, output, results, delta_g)
+      - docking (protein, job_id, smiles_hash, complex_hash, output, results)
+    """
+    databases = PrettyDict()
+
+    # Define a helper function to process a single database
+    def process_db(db_key: str, db_name, required_columns: list) -> tuple[str, object]:
+        """ensure that a db exists, and ensure that columns exist"""
+
+        db = _ensure_database(db_name)
+        db = _ensure_columns(database=db, required_columns=required_columns)
+
+        return db_key, db
+
+    # Mapping each database key to its parameters (name and required columns)
+    tasks = {
+        "ligands": (
+            DB_LIGANDS,
+            [dict(name=COL_LIGAND, type="file")],
+        ),
+        "proteins": (
+            DB_PROTEINS,
+            [dict(name=COL_PROTEIN, type="file")],
+        ),
+        "abfe": (
+            DB_ABFE,
+            [
+                dict(name=COL_PROTEIN, type="text"),
+                dict(name=COL_LIGAND1, type="text"),
+                dict(name=COL_COMPLEX_HASH, type="text"),
+                dict(name=COL_STEP, type="text"),
+                dict(name=COL_JOBID, type="text"),
+                dict(name=COL_CSV_OUTPUT, type="file"),
+                dict(name=COL_RESULT, type="file"),
+                dict(name=COL_DELTA_G, type="float"),
+            ],
+        ),
+        "rbfe": (
+            DB_RBFE,
+            [
+                dict(name=COL_PROTEIN, type="text"),
+                dict(name=COL_LIGAND1, type="text"),
+                dict(name=COL_LIGAND2, type="text"),
+                dict(name=COL_COMPLEX_HASH, type="text"),
+                dict(name=COL_STEP, type="text"),
+                dict(name=COL_JOBID, type="text"),
+                dict(name=COL_CSV_OUTPUT, type="file"),
+                dict(name=COL_RESULT, type="file"),
+                dict(name=COL_DELTA_DELTA_G, type="float"),
+            ],
+        ),
+        "docking": (
+            DB_DOCKING,
+            [
+                dict(name=COL_PROTEIN, type="text"),
+                dict(name=COL_JOBID, type="text"),
+                dict(name=COL_SMILES_HASH, type="text"),
+                dict(name=COL_COMPLEX_HASH, type="text"),
+                dict(name=COL_CSV_OUTPUT, type="file"),
+                dict(name=COL_RESULT, type="file"),
+            ],
+        ),
+    }
+
+    # Use a ThreadPoolExecutor for I/O-bound network calls
+    with concurrent.futures.ThreadPoolExecutor() as executor:
+        future_to_key = {
+            executor.submit(process_db, key, db_name, req_cols): key
+            for key, (db_name, req_cols) in tasks.items()
+        }
+
+        for future in concurrent.futures.as_completed(future_to_key):
+            key, db = future.result()
+            databases[key] = db
+
+    return databases
```

Comment:
note that this is parallelized, but it's still very slow. need to investigate why